core/modules/replica/tools/qserv-replica-job-delete-worker.cc

```diff
@@ -0,0 +1,199 @@
+/*
+ * LSST Data Management System
+ * Copyright 2018 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+/// qserv-replica-job-delete-worker.cc is a Controlller application
+/// for testing the corresponding job.
+
+// System headers
+#include <atomic>
+#include <iomanip>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+
+// Qserv headers
+#include "proto/replication.pb.h"
+#include "replica/Controller.h"
+#include "replica/ReplicaInfo.h"
+#include "replica/DeleteWorkerJob.h"
+#include "util/BlockPost.h"
+#include "util/CmdLineParser.h"
+
+using namespace lsst::qserv;
+
+namespace {
+
+// Command line parameters
+
+std::string worker;
+std::string configUrl;
+bool        permanentDelete;
+bool        progressReport;
+bool        errorReport;
+bool        chunkLocksReport;
+
+/// Run the test
+bool test() {
+
+    try {
+
+        ///////////////////////////////////////////////////////////////////////
+        // Start the controller in its own thread before injecting any requests
+        // Note that omFinish callbak which are activated upon a completion
+        // of the requsts will be run in that Controller's thread.
+
+        replica::ServiceProvider::Ptr const provider   = replica::ServiceProvider::create(configUrl);
+        replica::Controller::Ptr      const controller = replica::Controller::create(provider);
+
+        controller->run();
+
+        ////////////////////
+        // Start replication
+
+        std::atomic<bool> finished{false};
+        auto job = replica::DeleteWorkerJob::create(
+            worker,
+            permanentDelete,
+            controller,
+            std::string(),
+            [&finished] (replica::DeleteWorkerJob::Ptr job) {
+                finished = true;
+            }
+        );
+        job->start();
+
+        util::BlockPost blockPost(1000,2000);
+        while (not finished) {
+            blockPost.wait();
```

Comment:
why spin lock here? why not use a condition variable?

Reply:
Because I may need to track and report a progress of the (potentially long) operation. And I had this tracker in some (earlier) version of the application. Beside, the spin lock is the most trivial and easy to code and understand solution which has nearly 0 performance penalty on this application.